design-proposals/shadow-node.md

```diff
@@ -0,0 +1,157 @@
+The Heartbeat is high frequency, but it is not necessary to sync on to the Node (requires re-working node controller tp watch for the heartbeat annotation the shadowNode instead).
+With all the above doubling #writes is unlikely. 
+
+## Update/Rollback Compatibility
+
+virt-operator upgrade [order](https://github.com/kubevirt/kubevirt/blob/2e3a2a32d88a2e08c136c051c30754d1b4af423b/pkg/virt-operator/resource/apply/reconcile.go#L526-L641) is:
+```
+CRDs ==> RBACs ==> Daemonsets (=virt-handler) ==> Controller Deployments (=virt-controller) ==> API Deployments (=virt-API)
+```
+
+In order to handle backwards compatibility issues and minimize heartbeat timeouts, Virt-handler is going to continue writing to Node object.
+At some point during the upgrade, it would start being blocked due to the removal of the patch ability. In order to mitigate this and minimize heartbeat timeout during upgrade, the node patch will start ignoring RBAC "forbidden" error.
+Normally, new CRDs are blocked until the upgrade rollout is over. The shadowNode CRD is going to be excluded in order to avoid upgrade issues that could cause heartbeat timeout.
+
+Having said the above, this is a specific mention to each upgrade scenario - before and after the upgrade itself:
+
+### During upgrade rollout transient case scenarios:
+* New CRD; old RBAC; old virt-handler; old virt-controller; old virt-API:
+  * Old RBAC: no upgrade issue
+  * Old virt-handler: no upgrade issue
+  * Old virt-controllers: no upgrade issue
+  * Old Virt-API: no upgrade issue
+* New CRD, new RBAC; old virt-handler; old virt-controller; old virt-API:
+  * Old virt-handler: during upgrade, the old RBACs are kept as backup until upgrade is done. Hence, patching node should not fail so there is no upgrade issue.
+  * Old virt-controllers: no upgrade issue, virt controller RBAC is not changed due to this solution.
+  * Old Virt-API: no upgrade issue.
+* New CRD; new RBAC; new virt-handler; old virt-controller; old virt-API:
+  * new virt-handler: shadowNode is written to but it is not synced to the node yet. However, virt-handler keeps patching the node directly, so there is no upgrade issue.
+  * [same as before] Old controllers: no upgrade issue, virt controller RBAC is not changed.
+  * [same as before] Old Virt-API: no upgrade issue.
+* New CRD; new RBAC; new virt-handler; new virt-controller; old virt-API:
+  * New virt-handler: no upgrade issue. virt-handler patches both node and shadow node.
+  * New controllers: will start getting shadowNode requests once they are created. no upgrade issue.
+  * [same as before] Old Virt-API: no upgrade issue.
+
+### After upgrade rollout:
+Virt-handler will keep trying (and failing) to patch the node since the backup RBACs are now removed. Virt-handler will ignore these cases by ignoring RBAC "forbidden" errors. This behavior should be kept until we no longer support upgrades from the current release.
```

Comment:
This reads like you would never stop trying to patch. In case I did not misread this, let's suppose we have 100 nodes, we update the node every 10s (can't remember the default interval), and we may be able to remove this code path in 12 months, that means we would now patch and fail for 12 months  with 10 calls/second. So that is definitely not an option. Just think about metrics collection for instance which will record the failed call on virt-handler as well as on the API server.

The only way would be to stop patching when you get the "forbidden" error, but then we have again the issue, that we can't differentiate a config error from a real desire to no longer patch it.

Reply:
Indeed stop after first "forbidden" error is possible, but you are right that we can't differentiate a config error from a real desire to no longer patch it.